File: 268. Missing Number.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    // brute force: O(N) time & space, 0.53%
    int missingNumber(vector<int>& nums) {
        if(nums.empty()) return 0;
        set<int> number;
        for(int i = 0; i <= nums.size(); i++) number.insert(i);
        for(auto n: nums) number.erase(n);
        auto it = number.begin();
        return *it;
    }

    // sum and subtract: O(N) time, O(1)space, 54.63%, ok
    int missingNumber(vector<int>& nums) {
        if(nums.empty()) return 0;
        int sum = (1 + nums.size()) * nums.size() / 2;
        for(auto n: nums) sum -= n;
        return sum;
    }

    // XOR: O(N) time, O(1)space, 47.44%, ok
    int missingNumber(vector<int>& nums) {
        if(nums.empty()) return 0;
        int ret = 0, i = 1;
        for(auto n: nums){
            ret ^= n;
            ret ^= i++;
        }
        return ret;
    }
};

int main(){
    int test[7] = {2,1,3,5,6,0,7};
    vector<int> t(test, test + 7);
    Solution sol;
    cout << sol.missingNumber(t);
}

File: 268. Missing Number.py
class Solution(object):
    
    # sort: O(NlogN) time, no space
    def missingNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        nums.sort()
        if nums[0] != 0:
            return 0
        for i in xrange(1, len(nums)):
            if nums[i] != nums[i - 1] + 1:
                return nums[i] - 1
        return len(nums)
    
    # hash table: O(N) time & space
    def missingNumber(self, nums):
        nums = {num:1 for num in nums}
        for i in xrange(len(nums) + 1):
            if i not in nums:
                return i
    
    # sum and difference: O(N) time, O(1) space
    def missingNumber(self, nums):
        total = 0
        for i in xrange(len(nums)):
            total += i - nums[i]
        return total + len(nums)
    
    # xor: O(N) time, O(1) space
    def missingNumber(self, nums):
        ret = 0
        for i in xrange(len(nums)):
            ret ^= nums[i] ^ i
        return ret ^ len(nums)
            
